source/extensions/filters/http/cache/http_cache.h

```diff
@@ -0,0 +1,299 @@
+  // ensure that they can efficiently serve these ranges, and may merge and/or
+  // reorder ranges as appropriate, or may clear() response_ranges entirely.
+  std::vector<AdjustedByteRange> response_ranges;
+
+  // TODO(toddmgreer) Implement trailer support.
+  // True if the cached response has trailers.
+  bool has_trailers = false;
+};
+
+// Produces a hash of key that is consistent across restarts, architectures,
+// builds, and configurations. Caches that store persistent entries based on a
+// 64-bit hash should (but are not required to) use stableHashKey. Once this API
+// leaves alpha, any improvements to stableHashKey that would change its output
+// for existing callers is a breaking change.
+//
+// For non-persistent storage, use MessageUtil, which has no long-term stability
+// guarantees.
+//
+// When providing a cached response, Caches must ensure that the keys (and not
+// just their hashes) match.
+//
+// TODO(toddmgreer) Ensure that stability guarantees above are accurate.
+size_t stableHashKey(const Key& key);
+
+// LookupRequest holds everything about a request that's needed to look for a
+// response in a cache, to evaluate whether an entry from a cache is usable, and
+// to determine what ranges are needed.
+class LookupRequest {
+public:
+  using HeaderVector = std::vector<Http::HeaderEntry>;
+
+  // Prereq: request_headers's Path(), Scheme(), and Host() are nonnull.
+  LookupRequest(const Http::HeaderMap& request_headers, SystemTime timestamp);
+
+  // Caches may modify the key according to local needs, though care must be
+  // taken to ensure that meaningfully distinct responses have distinct keys.
+  const Key& key() const { return key_; }
+  Key& key() { return key_; }
+
+  // Returns the subset of this request's headers that are listed in
+  // envoy::config::filter::http::cache::v2alpha::Cache::allowed_vary_headers.
+  // Only needed if the cache implementation finds a variant response.
+  HeaderVector& vary_headers() { return vary_headers_; }
+  const HeaderVector& vary_headers() const { return vary_headers_; }
+
+  // Time when this LookupRequest was created (in response to an HTTP request).
+  SystemTime timestamp() const { return timestamp_; }
+
+  // WARNING: Incomplete--do not use in production (yet).
+  // Returns a LookupResult suitable for sending to the cache filter's
+  // LookupHeadersCallback. Specifically,
+  // - LookupResult::cache_entry_status is set according to HTTP cache
+  // validation logic.
+  // - LookupResult::headers takes ownership of response_headers.
+  // - LookupResult::content_length == content_length.
+  // - LookupResult::response_ranges entries are satisfiable (as documented
+  // there).
+  LookupResult makeLookupResult(Http::HeaderMapPtr&& response_headers,
+                                uint64_t content_length) const;
+
+private:
+  bool isFresh(const Http::HeaderMap& response_headers) const;
+  bool adjustByteRangeSet(std::vector<AdjustedByteRange>& response_ranges,
```

Comment:
I'd love to see some thorough unit-testing of this interesting and non-trivial method. Can you make this protected or public or friend a peer testing class to make that possible?

Alternatively could you factor out the math underlying this class to a static helper method or utility function?

Reply:
Done.